### ISSUE TITLE
Allow index files when checking a folder

### ISSUE THREAD

**Issue by dkarlovi** [lycheeverse/lychee#542]
_(note: this was previously discussed in https://github.com/lycheeverse/lychee/issues/452#issuecomment-1051025563)_

When deploying HTML files to some common platforms like Netlify or Gitlab Pages (or even on a typical Apache HTTP environment), you can rely on the platform to use an "index file", namely, for an URL like `https://example.com/foo`, it will serve `./public/foo/index.html` if such a file exists.

This means that, the heuristic should look like this:
1. is the "index file" feature enabled?
2. if so, does the URL naturally resolve to a folder?
3. if so, does the folder contain a file called `index.html` (or `index.htm`, maybe allow passing a list when enabling the feature)
4. if so, the URL is actually valid

The feature could be enabled as `lychee --allow-index-file index.html` (where the name is used in 3)

**Comment by mre** [comment]
Is this for checking files offline or online?

I'm guessing Apache would forward the request to the correct `index.html`/`index.html` file automatically with a redirect.

For offline checks, this could indeed be helpful. At the moment we only check if the directory exists where the `index.html` has to be in. The idea was that the web server would "do the right thing" if the folder existed. For example, if no `index.html` existed, it might show a directory listing instead if that is configured.
So in this case, even if the no `index.html` existed, this could still be a valid request. We'd only know by looking at the server config.

I like that your idea doesn't require knowledge about the server config and instead the user would be responsible for making the decision.

**Comment by dkarlovi** [comment]
> Is this for checking files offline or online?

Only offline, I think?

> I like that your idea doesn't require knowledge about the server config and instead the user would be responsible for making the decision.

Indeed, I actually have a use case where this should *not* work, as in, even if the `index.html` file exists, pointing to its parent repo should be treated as a failure (because the deployment target does *not* support the index file feature).

**Comment by lebensterben** [comment]
there're too much work for this feature request.

for netlify, you need to parse per-directory configuration and check "redirects". See https://docs.netlify.com/configure-builds/file-based-configuration/

for apache, you need DocumentRoot.

for gitlab we need to check "pages" job in .gitlab-ci.yml. specifically we need pages.artifects.paths to determine where index file is stored in a repository.

**Comment by dkarlovi** [comment]
@lebensterben I think there's a misunderstanding: you don't need to do any of that. You **let the user decide** if index files are supported or not for their use case with a runtime flag. If they say the index files are supported, lychee checks for them. If the user says they're not supported, lychee doesn't check for them.

If the user's wrong, it's up to them (in this case, me) to configure lychee correctly. There's no need to build any sort of heuristic if the user is able to define what's "correct".

**Comment by lebensterben** [comment]
nope, my point is you MUST have knowledge of server config.

For example, on netlify the index file may be named anything in per-directory configuration. How do you deal with that?

**Comment by mre** [comment]
The way I understood it is that you have to set the index file as a parameter.

```
lychee --allow-index-file foo.html
```

**Comment by dkarlovi** [comment]
Let the user run Lychee once per directory. If they choose to have such a complex setup, they could still do it, but it would be annoying (for them).

**Comment by mre** [comment]
This can now be (partially?) achieved thanks to our improved remap functionality from #1133.

E.g. 

```
lychee --remap 'example.com/([^/]*)$ example.com/$1/index.html' https://example.com/foo
```

This command will match URLs that do not contain a slash in the last segment of the path, implying they do not have a file extension like .html, and appends /index.html to the end of these URLs. Therefore, URLs like https://example.com/foo.html won't be affected by this command, as they already have .html at the end. But URLs like https://example.com/foo will be transformed to https://example.com/foo/index.html.

This matches with the documentation:

> Typically, a document called index.html will be served when a directory is requested without a file name being specified. 

@dkarlovi, would that work for you (or have worked in the past when you needed it)?
If so, we don't need the extra `--allow-index-file` option, and can fully rely on `--remap`.

If your use-case was more complex than that, we might need to switch to [fancy-regex](https://github.com/fancy-regex/fancy-regex), which supports negative lookahead, so you could do

```
lychee --remap 'example.com/((?!.html).)*$ example.com/$1/index.html' https://example.com/foo
```

...which would only match paths that don't end with `.html`. This would be even more precise, I guess, but I would hold off to add that, unless people need it.

**Comment by dkarlovi** [comment]
IIUC, your example will treat all URLs without an `.html` like they're actually `dir/index.html`? Would that work locally, with files and folders? If so, I don't see why it wouldn't be good enough. :thinking:

**Comment by dkarlovi** [comment]
Thinking more about this, it would likely improve UX if there was a super-common remap list shipped with Lychee and you could just enable it, without needing to copy/paste the expressions.

For example,
1. Lychee introduces a concept of a "named remap repository"
2. a few super common remaps are shipped by default, say `allow-index-file` creates the remap you've just described
3. you can then run `lychee --remap @allow-index-file` (just an example)
4. could also allow user to add their own in the config file, for example `lychee --remap @user-defined-remap`

**Comment by mre** [comment]
That's a creative idea!

One caveat would be, that we'd have to be super careful which remaps we ship by default. We probably don't want to change them too often, or we might break use-cases. Those remaps are generally very user-specific. E.g. in the above example, a user might redirect to `index.html` or `index.htm` according to their server config. The customizability is what makes them powerful.

If at all, then I'd opt for the separate repo and ask people to contribute. Then a user could copy them into their own config. On the other side, a user can already add rewrite expressions to their config, so I'm not sure if it's worth the additional overhead?

We could start by collecting these remaps to find out if there's demand for that feature. Since we can start with a text file, I'll create the repo right away, so we can start collecting. 

[Here it is!](https://github.com/lycheeverse/remaps)

It's a bit verbose right now, but we can tweak the format however we like. Everyone should feel welcome to contribute.
In the future, we can take it one step further and convert them into "well-known" remaps or something.

What do you think? 😃

**Comment by dkarlovi** [comment]
You're right, I didn't think of the `index.html` vs `index.htm` issue, it would indeed complicate things! :see_no_evil: 

Let's start with your suggestion, it indeed makes sense to first see if there's a possibility of some remaps actually being "super common", if so, it should be easy enough to have Lychee ship them. :+1:

**Comment by katrinafyi** [comment]
It looks like this will finally be implemented in https://github.com/lycheeverse/lychee/pull/1752

edit: https://github.com/lycheeverse/lychee/pull/1777
